coil-network/src/commonMain/kotlin/coil3/fetch/NetworkFetcher.kt

```diff
@@ -0,0 +1,287 @@
+            snapshot?.closeQuietly()
+            return null
+        }
+
+        // Open a new editor.
+        val editor = if (snapshot != null) {
+            snapshot.closeAndOpenEditor()
+        } else {
+            diskCache.value?.openEditor(diskCacheKey)
+        }
+
+        // Return null if we're unable to write to this entry.
+        if (editor == null) return null
+
+        try {
+            // Write the response to the disk cache.
+            if (networkResponse.status == HttpStatusCode.NotModified && cacheResponse != null) {
+                // Combine and write the updated cache headers and discard the body.
+                fileSystem.write(editor.metadata) {
+                    CacheResponse(
+                        response = networkResponse,
+                        headers = Headers.build {
+                            appendAll(networkResponse.headers)
+                            appendAllIfNameAbsent(cacheResponse.responseHeaders)
+                        },
+                    ).writeTo(this)
+                }
+                networkResponse.discardRemaining()
+            } else {
+                // Write the response's cache headers and body.
+                fileSystem.write(editor.metadata) {
+                    CacheResponse(networkResponse).writeTo(this)
+                }
+                networkResponseBody.writeTo(fileSystem, editor.data)
+            }
+            return editor.commitAndOpenSnapshot()
+        } catch (e: Exception) {
+            editor.abortQuietly()
+            throw e
+        }
+    }
+
+    private fun newRequest(): HttpRequestBuilder {
+        val request = HttpRequestBuilder()
+        request.method = options.httpMethod
+        request.url.takeFrom(url)
+        request.headers.appendAll(options.httpHeaders)
+
+        val diskRead = options.diskCachePolicy.readEnabled
+        val networkRead = options.networkCachePolicy.readEnabled
+        when {
+            !networkRead && diskRead -> {
+                request.header(CACHE_CONTROL, "only-if-cached, max-stale=2147483647")
+            }
+
+            networkRead && !diskRead -> if (options.diskCachePolicy.writeEnabled) {
+                request.header(CACHE_CONTROL, "no-cache")
+            } else {
+                request.header(CACHE_CONTROL, "no-cache, no-store")
+            }
+
+            !networkRead && !diskRead -> {
+                // This causes the request to fail with a 504 Unsatisfiable Request.
+                request.header(CACHE_CONTROL, "no-cache, only-if-cached")
+            }
+        }
+
+        return request
+    }
+
+    private suspend fun <T> executeNetworkRequest(
+        request: HttpRequestBuilder,
+        block: suspend (HttpResponse) -> T,
+    ): T {
+        // Prevent executing requests on the main thread that could block due to a
+        // networking operation.
+        if (options.networkCachePolicy.readEnabled) {
+            assertNotOnMainThread()
+        }
```

Comment:
I don't think this is still needed, ktor is not okhttp, it's fully async, no way to do network on MainThread unless engine author is a fool who use Dispatcher.Main to do socket select..
But we still need take care of disk cache writing to ensure it not happen on MainThread

Reply:
@VDostoyevskiy Some users previously requested the ability to force loading the image from the disk cache on the main thread. It's definitely not recommended, but can be supported by requiring the user to set the `networkCachePolicy = CachePolicy.DISABLED` if they want to use the main thread.